x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/UriUtils.java

```diff
@@ -42,42 +56,211 @@ public static URI parseURI(String connectionString, URI defaultURI) {
+            }
+            if (uri.getRawPath() != null) {
+                sb.append(uri.getRawPath());
+            }
+            if (uri.getQuery() != null) {
+                sb.append('?');
+                // redact with the null character; this will later allow safe reinsertion of any character removed from the URI to make
+                // it parsable
+                sb.append(redactCredentialsInRawUriQuery(uri.getRawQuery(), '\0'));
+            }
+            if (uri.getRawFragment() != null) {
+                sb.append('#');
+                sb.append(uri.getRawFragment());
+            }
+
+            // reinsert any removed character back into the URI: if the reinsertion should be made between null characters, replace its
+            // value with a null character, since it's part of the credential value
+            Collections.reverse(faults);
+            for (Map.Entry<Integer, Character> e : faults) {
+                int idx = e.getKey();
+                if (idx >= sb.length()) {
+                    sb.append(e.getValue());
+                } else {
+                    sb.insert(idx,
+                        (sb.charAt(idx) == '\0' && (idx + 1 >= sb.length() || sb.charAt(idx + 1) == '\0')) ? '\0' : e.getValue());
+                }
+            }
+
+            StringBuilder ret = new StringBuilder();
+            sb.chars().forEach(x -> ret.append(x == '\0' ? REDACTION_CHAR : (char) x));
+
+            return ret.toString();
+        }
+
+        private static String redactCredentialsInURLString(String urlString) {
+            List<Map.Entry<Integer, Character>> faults = new ArrayList<>();
+
+            boolean hasPort = false;
+            for (StringBuilder sb = new StringBuilder(urlString); sb.length() > 0; ) {
+                try {
+                    // parse as URL; ex. `http://ho~st` parses as URI, but with unparsable authority
+                    URI uri = new URI(sb.toString()).parseServerAuthority();
+                    return editURI(uri, faults, hasPort);
+                } catch (URISyntaxException use) {
+                    int idx = use.getIndex();
+                    if (idx < 0 || idx >= sb.length()) {
+                        break; // not a faulty character-related error
+                    }
+                    if (use.getReason().equals("Illegal character in port number")) {
+                        // if entire port part is broken (ex. `localhost:noDigit`), the trailing `:` will be lost in the resulting URI
+                        hasPort = true;
+                    }
+                    faults.add(new AbstractMap.SimpleImmutableEntry<>(use.getIndex(), sb.charAt(idx)));
+                    sb.deleteCharAt(idx);
+                }
+            }
+            return null;
+        }
+
+        public static String redactCredentialsInConnectionString(String connectionString) {
+            if (connectionString.startsWith(HTTP_PREFIX.toUpperCase(Locale.ROOT))
```

Comment:
> Do we have a good/strong argument in favor of "deactivating" the redaction if URLs when the schema is uppercase and activating it when it's lowercase?

Just a small clarification: redaction is currently active in all cases, except an all-uppercase schema.

As @astefan pointed out, this can be an issue if the user provides it like that unknowingly initially (potentially generated all-uppercase by software). Hopefully this happens very rarely, but we can choose something else if we are concerned about it (like some `HttP` camelcase or even a completely different schema, though that might be a bit confusing).

> Wouldn't a URL parameter do the job, similar to `validate.properties`?

The trick here would be to reliably recognise that parameter in case the URI itself is broken. We could also do a simple string scanning (`lower_case_uri.contains("redaction=false")`) and hope for no false positives - should be indeed very unlikely -, but the downside of that is that the user would potentially no longer get the same error message if offsets change (which is however a possibility with current mechanism as well, in case the user didn't provide a schema in the first place), or they could introduce new errors.

Reply:
We would default the `redaction` parameter to `true` and if the the URL is broken and the parameter is not recognizable the URL sensitive data wouldn't leak anyway. If the user is interested in disabling the redaction step, then will look into the URL correctness anyway.

Reply:
Yep, that could also work.
Myself I'm still biased towards the existing implementation as it's slightly simpler and doesn't require an extra param, but "close to the fence", so if anyone would have a swaying opinion I'll change to the above (or potentially new) suggestion. CC: @matriv @costin

Reply:
Personally I'd vote to keep it as is instead of introducing a new parameter. Do we have any experience/demand in the past for this uppercase URIs? To me seems like a super edge case and wouldn't invest too much to have the perfect handling for it.